### ISSUE TITLE
arm/compute.VirtualMachinesClient.Get fails with 503 Service Unavailable

### ISSUE THREAD

**Issue by axw** [Azure/azure-sdk-for-go#231]
Mostly I'm using the List method, but I'd like to optimise a case where we're only interested in querying a single instance. To that end, I've attempted to use the Get method, only to find it doesn't seem to work.

The Get method has a mysterious "expand" parameter, which AFAICT, does not correspond to anything documented against the compute REST API. The only reference I could find for $expand (which is how it's serialised) is to the AD Graph API. Anyway, I've set it to "". If I comment out https://github.com/Azure/azure-sdk-for-go/blob/master/arm/compute/virtualmachines.go#L403, Get works just fine, so $expand does appear to be at fault.

```
autorest:DoErrorUnlessStatusCode GET https://management.azure.com/subscriptions/6f0e3b87-f601-40fd-9727-76d05563fedd/resourceGroups/juju-azure-next-environment-14e690dc-d5c9-4958-8d6a-06004681e7eb/providers/Microsoft.Compute/virtualMachines/machine-0?%24expand=&api-version=2015-06-15 failed with 503 Service Unavailable
```

**Comment by axw** [comment]
I haven't tested with VirtualMachineExtensionsClient.Get, but it also passes $expand.

**Comment by brendandixon** [comment]
@axw $expand is part of the OData support used across the APIs (see http://www.odata.org). However, those values are optional. The SDK has a bug in that it is failing to distinguish between required and optional parameters. I will try to get a build out that addresses this soon.

**Comment by axw** [comment]
@brendandixon Thanks, and sorry, I now see there's a doc-comment in the Go code about it accepting a value of "instanceView". It's a shame I can't use this when listing all VMs, as I'd like to be able to get the power state without having to query VMs individually.

**Comment by brendandixon** [comment]
@axw One of the features missing from the Go SDK is OData support. Many of the APIs accept those parameters. But, sadly, it does not appear that ARM Listing of VMs is one of those. That's odd in my mind. Let me look into this more.

**Comment by axw** [comment]
@brendandixon Thanks. I've pulled the latest azure-sdk-for-go, and the models seemed to have changed a fair bit. Some of the field types don't match up to enum types any more. For example, compute.HardwareProfile.VMSize is now a string, whereas previously it was VirtualMachineSizeTypes. Is this intentional?

**Comment by brendandixon** [comment]
@axw Let me take a closer look.

Enums in Swagger are tricky. Core Swagger defines enumerations, but does not provide a means to clearly type them. That is, it supports restricting a field to a set of values, but lacks some of that necessary to expose those values as a set (e.g., an enumeration). The Autorest work, which is generating the Go ARM packages, relies on an Microsoft extension to Swagger -- the x-ms-enum field -- to convey more details so that we can promote enumerations to full types.

It could be the Swagger changed or that a change I made brought about what you're seeing.

Are only enumerated types affected? Or have you seen other differences as well? I did change the maps used to represent an object type to be "map[string]interface{}" from "map[string]string". The latter is actually incorrect. But I did not intend any other type changes to occur.

**Comment by axw** [comment]
@brendandixon Apart from losing types on enum fields, and map[string]\*string -> map[string]interface{} here's the changes I've needed to make in Juju.

**arm/network**
ManagementClient -> ResourceProviderClient
SecurityRuleProtocolTCP -> TCP
SecurityRuleProtocolUDP -> UDP

**arm/resources**
Group -> ResourceGroup (client is still called GroupsClient)

------------

The change in name of network.ManagementClient makes it inconsistent with each of the other providers (at least the ones we use). TCP/UDP seems fine, but might collide with other enums? And ResourceGroup is inconsistent with the client name. Fairly cosmetic issues, though.

**Comment by brendandixon** [comment]
@axw Thank you much for this. I'll look into these.

I did make changes that could, and appears did, affect client names. Prior to this latest change, the generator hard-coded "ManagementClient." That is actually incorrect and does not follow how the other generators worked. I changed the code to correctly derive names, but to do so in a way not likely to generate changes (which I could see via the diffs). It appears there is a case that escaped my notice.

In any case, let me look at this more and I'll respond. Thank you for your patience!

**Comment by brendandixon** [comment]
@axw I poked around a bit and found a couple of interesting items. First, I made a change to the generator to address one problem, but its effect was to lose the enumeration type names within structures. I can and will fix that. Second, the Network swagger changed. But it was an odd change. Their is a new Swagger that uses the correct client name (i.e., ManagementClient), but the old Swagger (which the generator still used) was changed to use the name you're seeing.

I will generate an update to address both of these issues.
